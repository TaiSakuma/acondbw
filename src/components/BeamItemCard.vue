<template>
  <div class="beam-item-card" style="position: relative;">
    <v-card outlined hover :loading="state == State.LOADING" style="max-width: 980px;">
      <v-card-text v-if="state == State.LOADING" class="pa-2">loading...</v-card-text>
      <v-card-text v-else-if="state == State.ERROR">Error: cannot load data</v-card-text>
      <div v-else-if="state == State.LOADED" @click="$emit('expand')" style="cursor: default;">
        <v-container fluid class="pa-0">
          <v-row class="ma-0 px-0">
            <v-col order="1" cols="10" md="3" class="py-0">
              <div class="caption grey--text">Name</div>
              <div class="font-weight-medium primary--text">
                <span @click.stop>
                  <router-link :to="'/beams/item/' + beam.name" v-text="beam.name"></router-link>
                </span>
              </div>
            </v-col>
            <v-col order="3" cols="12" md="7" class="py-0">
              <div class="caption grey--text">Path</div>
              <div v-text="beam.path"></div>
            </v-col>
            <v-col order="2" order-md="4" cols="2" align-self="end" class="py-0">
              <v-row align="start" justify="end" class="px-1 py-0">
                <div v-if="collapsible">
                  <v-tooltip bottom open-delay="800">
                    <template v-slot:activator="{ on }">
                      <v-btn
                        icon
                        @click.stop="
                      collapsed ? $emit('expand') : $emit('collapse')
                    "
                        v-on="on"
                      >
                        <v-icon>
                          {{
                          collapsed ? "mdi-chevron-down" : "mdi-chevron-up"
                          }}
                        </v-icon>
                      </v-btn>
                    </template>
                    <span>{{ collapsed ? "Expand" : "Collapse" }}</span>
                  </v-tooltip>
                </div>
                <span @click.stop>
                  <v-menu left bottom offset-y v-model="menu" :close-on-content-click="false">
                    <template v-slot:activator="{ on: menu }">
                      <v-btn icon v-on="{ ...menu }">
                        <v-icon>mdi-dots-vertical</v-icon>
                      </v-btn>
                    </template>
                    <v-list dense>
                      <v-dialog v-model="editDialog" persistent max-width="600">
                        <template v-slot:activator="{ on: editDialog }">
                          <v-list-item disabled v-on="{ ...editDialog }">
                            <v-list-item-icon>
                              <v-icon disabled>mdi-pencil</v-icon>
                            </v-list-item-icon>
                            <v-list-item-content>
                              <v-list-item-title>Edit</v-list-item-title>
                            </v-list-item-content>
                          </v-list-item>
                        </template>
                      </v-dialog>
                      <v-dialog v-model="deleteDialog" max-width="600">
                        <template v-slot:activator="{ on: deleteDialog }">
                          <v-list-item disabled v-on="{ ...deleteDialog }">
                            <v-list-item-icon>
                              <v-icon disabled>mdi-delete</v-icon>
                            </v-list-item-icon>
                            <v-list-item-content>
                              <v-list-item-title>Delete</v-list-item-title>
                            </v-list-item-content>
                          </v-list-item>
                        </template>
                      </v-dialog>
                    </v-list>
                  </v-menu>
                </span>
              </v-row>
            </v-col>
          </v-row>
          <v-expand-transition>
            <v-row class="mx-0 mb-3 px-0 collapsible" v-show="!(collapsible && collapsed)">
              <v-col cols="12" md="8" offset-md="3" class="py-0">
                <div class="caption grey--text">Map</div>
                <div v-if="beam.map">
                  <router-link :to="'/maps/item/' + beam.map.name" v-text="beam.map.name"></router-link>
                </div>
              </v-col>
              <v-col cols="12" class="py-0">
                <v-row>
                  <v-col order="2" order-md="0" cols="12" md="3" align-self="end" class="py-0">
                    <span class="grey--text" style="font-size: 65%;">Data ID: {{ dataId }}</span>
                  </v-col>
              <v-col order="1" cols="12" md="8" class="py-0">
                <div class="caption grey--text">Parent Beam</div>
                <div v-if="beam.parentBeam">
                  <router-link
                    :to="'/beams/item/' + beam.parentBeam.name"
                    v-text="beam.parentBeam.name"
                  ></router-link>
                </div>
              </v-col>
                </v-row>
              </v-col>
            </v-row>
          </v-expand-transition>
        </v-container>
      </div>
      <v-card-text v-else>Nothing to show here.</v-card-text>
    </v-card>
    <DevToolLoadingStateOverridingMenu @state="devtoolState = $event"></DevToolLoadingStateOverridingMenu>
  </div>
</template>

<script>
import { defaultDataIdFromObject } from "apollo-cache-inmemory";

import BEAM from "@/graphql/Beam.gql";

import State from "@/utils/LoadingState.js";
import DevToolLoadingStateOverridingMenu from "@/components/DevToolLoadingStateOverridingMenu";

export default {
  name: "BeamItemCard",
  components: {
    DevToolLoadingStateOverridingMenu
},
  props: {
    beamId: { default: null }, // beam.beamId not beam.id
    collapsed: { default: false },
    collapsible: { default: false }
  },
  data() {
    return {
      menu: false,
      editDialog: false,
      deleteDialog: false,
      beam: null,
      error: null,
      devtoolState: null,
      State: State
    };
  },
  computed: {
    state() {
      if (this.devtoolState) {
        return this.devtoolState;
      }

      if (this.loading) {
        return State.LOADING;
      } else if (this.error) {
        return State.ERROR;
      } else if (this.beam) {
        return State.LOADED;
      } else {
        return State.NONE;
      }
    },
    loading() {
      return this.$apollo.queries.beam.loading;
    },
    dataId: function() {
      return defaultDataIdFromObject(this.beam);
    }
  },
  apollo: {
    beam: {
      query: BEAM,
      variables() {
        return {
          beamId: this.beamId
        };
      },
      result(result) {
        this.error = null;
        if (result.error) {
          this.error = true;
        }
      }
    }
  }
};
</script>
