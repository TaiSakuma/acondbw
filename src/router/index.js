import Vue from "vue";
import VueRouter from "vue-router";
import Home from "@/views/Home.vue";
import ProductTop from "@/views/ProductTop.vue";

import ALL_MAPS from "@/graphql/AllMaps.gql";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: "home",
    component: Home,
  },
  {
    path: "/about",
    name: "about",
    component: () =>
      import(/* webpackChunkName: "about" */ "@/views/About.vue"),
  },
  {
    path: "/simulations",
    component: ProductTop,
    props: {
      title: "Simulations",
      icon: "mdi-creation",
      itemPageName: "SimulationItem",
    },
    children: [
      {
        path: "",
        name: "SimulationList",
        component: () =>
          import(
            /* webpackChunkName: "simulations" */ "@/views/SimulationList.vue"
          ),
      },
      {
        path: "item/:name",
        name: "SimulationItem",
        component: () =>
          import(
            /* webpackChunkName: "simulations" */ "@/views/SimulationItem.vue"
          ),
      },
    ],
  },
  {
    path: "/maps",
    component: ProductTop,
    props: { title: "Maps", icon: "map", itemPageName: "MapItem" },
    children: [
      {
        path: "",
        name: "MapList",
        component: () =>
          import(/* webpackChunkName: "maps" */ "@/views/MapList.vue"),
        props: {
          productTypeName: { singular: "map", plural: "maps" },
          query: ALL_MAPS,
          queryName: "allMaps",
          productIdFieldName: "mapId",
          productItemCard: "MapItemCard",
          productAddForm: "MapAddForm",
          disableAdd: process.env.VUE_APP_ACONDBW_MAP_MUTATION_DIALOG != "true",
        },
      },
      {
        path: "item/:name",
        name: "MapItem",
        component: () =>
          import(/* webpackChunkName: "maps" */ "@/views/MapItem.vue"),
      },
    ],
  },
  {
    path: "/beams",
    component: ProductTop,
    props: {
      title: "Beams",
      icon: "mdi-spotlight-beam",
      itemPageName: "BeamItem",
    },
    children: [
      {
        path: "",
        name: "BeamList",
        component: () =>
          import(/* webpackChunkName: "beams" */ "@/views/BeamList.vue"),
      },
      {
        path: "item/:name",
        name: "BeamItem",
        component: () =>
          import(/* webpackChunkName: "beams" */ "@/views/BeamItem.vue"),
      },
    ],
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes,
});

export default router;
