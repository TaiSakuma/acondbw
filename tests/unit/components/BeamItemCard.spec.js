import Vue from "vue";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import { mount, createLocalVue } from "@vue/test-utils";

import BeamItemCard from "@/components/BeamItemCard.vue";
import router from "@/router";

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("BeamItemCard.vue", () => {
  let localVue;

  function createWrapper({ loading = false, propsData } = {}) {
    let wrapper = mount(BeamItemCard, {
      localVue,
      router,
      mocks: {
        $apollo: {
          queries: {
            beam: {
              loading: loading
            }
          }
        }
      },
      propsData: {
        beamName: "20200207",
        ...propsData
      }
    });

    return wrapper;
  }

  const beam = {
    beamId: "1150",
    name: "20200207",
    path: "BEAM_DEPOT/Beams/20200207",
    map: {
      name: "lat20200201"
    },
    parentBeam: {
      name: "20200123"
    }
  };

  beforeEach(function() {
    localVue = createLocalVue();
  });

  it("loading", async () => {
    const loading = true;
    const wrapper = createWrapper({ loading });
    await Vue.nextTick();
    expect(wrapper.text()).toContain("loading");
  });

  it("error", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      error: true
    });
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Error: cannot load data");
  });

  it("none", async () => {
    const wrapper = createWrapper();
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Nothing to show here");
  });

  it("match snapshot", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      beam: beam
    });
    await Vue.nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it.each([
    [true, true],
    [true, false],
    [false, true],
    [false, false]
  ])(
    "match snapshot - {collapsible: %p, collapsed: %p}",
    async (collapsible, collapsed) => {
      const wrapper = createWrapper({
        propsData: { collapsed: collapsed, collapsible: collapsible }
      });
      wrapper.setData({
        beam: beam
      });
      await Vue.nextTick();
      expect(wrapper.html()).toMatchSnapshot();
    }
  );
});
