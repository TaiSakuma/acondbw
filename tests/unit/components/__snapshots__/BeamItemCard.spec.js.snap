// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeamItemCard.vue match snapshot - {collapsible: false, collapsed: false} 1`] = `
<div class="beamitemcard">
  <div class="v-card v-card--hover v-card--outlined v-sheet theme--light" style="max-width: 980px;">
    <div style="cursor: default;">
      <div class="layout ma-0 px-3 row wrap">
        <div class="flex xs12 md3">
          <div class="caption grey--text">Name</div>
          <div class="font-weight-medium primary--text"><span><a href="/beams/item/20200207" class="">20200207</a></span></div>
        </div>
        <div class="flex xs11 md8">
          <div class="caption grey--text">Path</div>
          <div>BEAM_DEPOT/Beams/20200207</div>
        </div>
        <div class="flex xs1 md1 align-self-end">
          <!---->
        </div>
      </div>
      <div class="layout mx-0 mb-3 px-3 row wrap">
        <div class="flex xs12 md8 offset-md-3">
          <div class="caption grey--text">Map</div>
          <div><a href="/maps/item/lat20200201" class="">lat20200201</a></div>
        </div>
        <div class="flex xs12 md4 offset-md-3">
          <div class="caption grey--text">Parent Beam</div>
          <div></div>
          <div><a href="/beams/item/20200123" class="">20200123</a></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BeamItemCard.vue match snapshot - {collapsible: false, collapsed: true} 1`] = `
<div class="beamitemcard">
  <div class="v-card v-card--hover v-card--outlined v-sheet theme--light" style="max-width: 980px;">
    <div style="cursor: default;">
      <div class="layout ma-0 px-3 row wrap">
        <div class="flex xs12 md3">
          <div class="caption grey--text">Name</div>
          <div class="font-weight-medium primary--text"><span><a href="/beams/item/20200207" class="">20200207</a></span></div>
        </div>
        <div class="flex xs11 md8">
          <div class="caption grey--text">Path</div>
          <div>BEAM_DEPOT/Beams/20200207</div>
        </div>
        <div class="flex xs1 md1 align-self-end">
          <!---->
        </div>
      </div>
      <div class="layout mx-0 mb-3 px-3 row wrap">
        <div class="flex xs12 md8 offset-md-3">
          <div class="caption grey--text">Map</div>
          <div><a href="/maps/item/lat20200201" class="">lat20200201</a></div>
        </div>
        <div class="flex xs12 md4 offset-md-3">
          <div class="caption grey--text">Parent Beam</div>
          <div></div>
          <div><a href="/beams/item/20200123" class="">20200123</a></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BeamItemCard.vue match snapshot - {collapsible: true, collapsed: false} 1`] = `
<div class="beamitemcard">
  <div class="v-card v-card--hover v-card--outlined v-sheet theme--light" style="max-width: 980px;">
    <div style="cursor: default;">
      <div class="layout ma-0 px-3 row wrap">
        <div class="flex xs12 md3">
          <div class="caption grey--text">Name</div>
          <div class="font-weight-medium primary--text"><span><a href="/beams/item/20200207" class="">20200207</a></span></div>
        </div>
        <div class="flex xs11 md8">
          <div class="caption grey--text">Path</div>
          <div>BEAM_DEPOT/Beams/20200207</div>
        </div>
        <div class="flex xs1 md1 align-self-end">
          <div>
            <div class="layout justify-end"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-up theme--light"></i></span></button><span class="v-tooltip v-tooltip--bottom"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"></div></span></div>
          </div>
        </div>
      </div>
      <div class="layout mx-0 mb-3 px-3 row wrap">
        <div class="flex xs12 md8 offset-md-3">
          <div class="caption grey--text">Map</div>
          <div><a href="/maps/item/lat20200201" class="">lat20200201</a></div>
        </div>
        <div class="flex xs12 md4 offset-md-3">
          <div class="caption grey--text">Parent Beam</div>
          <div></div>
          <div><a href="/beams/item/20200123" class="">20200123</a></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BeamItemCard.vue match snapshot - {collapsible: true, collapsed: true} 1`] = `
<div class="beamitemcard">
  <div class="v-card v-card--hover v-card--outlined v-sheet theme--light" style="max-width: 980px;">
    <div style="cursor: default;">
      <div class="layout ma-0 px-3 row wrap">
        <div class="flex xs12 md3">
          <div class="caption grey--text">Name</div>
          <div class="font-weight-medium primary--text"><span><a href="/beams/item/20200207" class="">20200207</a></span></div>
        </div>
        <div class="flex xs11 md8">
          <div class="caption grey--text">Path</div>
          <div>BEAM_DEPOT/Beams/20200207</div>
        </div>
        <div class="flex xs1 md1 align-self-end">
          <div>
            <div class="layout justify-end"><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-down theme--light"></i></span></button><span class="v-tooltip v-tooltip--bottom"><div class="v-tooltip__content" style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"></div></span></div>
          </div>
        </div>
      </div>
      <div class="layout mx-0 mb-3 px-3 row wrap" style="display: none;">
        <div class="flex xs12 md8 offset-md-3">
          <div class="caption grey--text">Map</div>
          <div><a href="/maps/item/lat20200201" class="">lat20200201</a></div>
        </div>
        <div class="flex xs12 md4 offset-md-3">
          <div class="caption grey--text">Parent Beam</div>
          <div></div>
          <div><a href="/beams/item/20200123" class="">20200123</a></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BeamItemCard.vue match snapshot 1`] = `
<div class="beamitemcard">
  <div class="v-card v-card--hover v-card--outlined v-sheet theme--light" style="max-width: 980px;">
    <div style="cursor: default;">
      <div class="layout ma-0 px-3 row wrap">
        <div class="flex xs12 md3">
          <div class="caption grey--text">Name</div>
          <div class="font-weight-medium primary--text"><span><a href="/beams/item/20200207" class="">20200207</a></span></div>
        </div>
        <div class="flex xs11 md8">
          <div class="caption grey--text">Path</div>
          <div>BEAM_DEPOT/Beams/20200207</div>
        </div>
        <div class="flex xs1 md1 align-self-end">
          <!---->
        </div>
      </div>
      <div class="layout mx-0 mb-3 px-3 row wrap">
        <div class="flex xs12 md8 offset-md-3">
          <div class="caption grey--text">Map</div>
          <div><a href="/maps/item/lat20200201" class="">lat20200201</a></div>
        </div>
        <div class="flex xs12 md4 offset-md-3">
          <div class="caption grey--text">Parent Beam</div>
          <div></div>
          <div><a href="/beams/item/20200123" class="">20200123</a></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
